File: python/src/agent_framework/core/permission_policy.py
from __future__ import annotations

import re
from dataclasses import dataclass

from agent_framework.core.domain import PermissionDecision


@dataclass
class PermissionRule:
    tool_name: str
    decision: PermissionDecision
    pattern: re.Pattern[str] | None = None


DEFAULT_CODING_POLICY_RULES = [
    PermissionRule("shell", "deny", re.compile(r"rm\s+-rf\s+/")),
    PermissionRule("shell", "ask", re.compile(r"sudo\s+")),
    PermissionRule("write_file", "allow"),
    PermissionRule("read_file", "allow"),
    PermissionRule("grep", "allow"),
    PermissionRule("glob", "allow"),
]


class DefaultPermissionPolicy:
    def __init__(self, rules: list[PermissionRule] | None = None) -> None:
        self._rules = rules or DEFAULT_CODING_POLICY_RULES

    async def check(self, tool_name: str, arguments: dict, session_id: str) -> PermissionDecision:
        del session_id
        for rule in self._rules:
            if rule.tool_name not in ("*", tool_name):
                continue
            if rule.pattern and tool_name == "shell":
                command = str(arguments.get("command", ""))
                if rule.pattern.search(command):
                    return rule.decision
                continue
            if rule.tool_name in (tool_name, "*"):
                return rule.decision
        return "allow"

File: typescript/packages/core/src/permission-policy.ts
import type { PermissionCheckRequest, PermissionDecision, PermissionPolicy } from './types/contracts.js';

export interface PermissionRule {
  toolName: string | '*';
  pattern?: RegExp;
  decision: PermissionDecision;
}

export class DefaultPermissionPolicy implements PermissionPolicy {
  constructor(private readonly rules: PermissionRule[] = []) {}

  async check(request: PermissionCheckRequest): Promise<PermissionDecision> {
    for (const rule of this.rules) {
      if (rule.toolName !== '*' && rule.toolName !== request.toolName) {
        continue;
      }
      if (rule.pattern && request.toolName === 'shell') {
        const command = String(request.arguments.command ?? '');
        if (rule.pattern.test(command)) {
          return rule.decision;
        }
        continue;
      }
      if (rule.toolName === request.toolName || rule.toolName === '*') {
        return rule.decision;
      }
    }
    return 'allow';
  }
}

export const DEFAULT_CODING_POLICY_RULES: PermissionRule[] = [
  { toolName: 'shell', pattern: /rm\s+-rf\s+\//, decision: 'deny' },
  { toolName: 'shell', pattern: /sudo\s+/, decision: 'ask' },
  { toolName: 'write_file', decision: 'allow' },
  { toolName: 'read_file', decision: 'allow' },
  { toolName: 'grep', decision: 'allow' },
  { toolName: 'glob', decision: 'allow' },
];
